src/ucp/core/ucp_am.c

```diff
@@ -1784,6 +1784,22 @@ const ucp_request_send_proto_t ucp_am_reply_proto = {
 void ucp_am_proto_request_zcopy_reset(ucp_request_t *request)
 {
     ucs_assert(request->send.msg_proto.am.header.reg_desc != NULL);
+
+    /* If user header is not guaranteed to be valid, 
+     * use mpool buffer for storing the user header.
+     */
+    if ((request->send.msg_proto.am.flags & UCP_AM_SEND_FLAG_COPY_HEADER) &&
+        !(request->flags & UCP_REQUEST_FLAG_USER_HEADER_COPIED) &&
```

Comment:
we are in zcopy protocol here.
seems the only way UCP_REQUEST_FLAG_USER_HEADER_COPIED could be set is if we "switched" from zcopy to zcopy?
maybe need to release request->send.msg_proto.am.header.ptr in this case?